subprojects/core/src/main/java/org/gradle/initialization/StartParameterBuildOptions.java

```diff
@@ -408,4 +413,78 @@ public void applyTo(StartParameterInternal settings, Origin origin) {
             settings.setExportKeys(true);
         }
     }
+
+    public static class ConfigurationCacheOption extends BooleanBuildOption<StartParameterInternal> {
+
+        public static final String PROPERTY_NAME = "org.gradle.unsafe.configuration-cache";
+        public static final String LONG_OPTION = "configuration-cache";
+
+        public ConfigurationCacheOption() {
+            super(PROPERTY_NAME, BooleanCommandLineOptionConfiguration.create(
+                LONG_OPTION,
+                "Enables the configuration cache. Gradle will try to reuse the build configuration from previous builds.",
+                "Disables the configuration cache."
+            ).incubating());
+        }
+
+        @Override
+        public void applyTo(boolean value, StartParameterInternal settings, Origin origin) {
+            settings.setConfigurationCacheEnabled(value);
+        }
+    }
+
+    public static class ConfigurationCacheFailOnProblemsOption extends BooleanBuildOption<StartParameterInternal> {
+
+        public static final String PROPERTY_NAME = "org.gradle.unsafe.configuration-cache.fail-on-problems";
+        public static final String LONG_OPTION = "configuration-cache-fail-on-problems";
+
+        public ConfigurationCacheFailOnProblemsOption() {
+            super(PROPERTY_NAME, BooleanCommandLineOptionConfiguration.create(
+                LONG_OPTION,
+                "Lets the configuration cache fail on problems. This is the default.",
+                "Lets the configuration cache do not fail on problems."
+            ).incubating());
+        }
+
+        @Override
+        public void applyTo(boolean value, StartParameterInternal settings, Origin origin) {
+            settings.setConfigurationCacheFailOnProblems(value);
+        }
+    }
+
+    public static class ConfigurationCacheMaxProblemsOption extends IntegerBuildOption<StartParameterInternal> {
+
+        public static final String PROPERTY_NAME = "org.gradle.unsafe.configuration-cache.max-problems";
+        public static final String LONG_OPTION = "configuration-cache-max-problems";
+
+        public ConfigurationCacheMaxProblemsOption() {
+            super(PROPERTY_NAME, CommandLineOptionConfiguration.create(
+                LONG_OPTION,
+                "Sets the maximum configuration cache problems."
+            ).incubating());
+        }
+
+        @Override
+        public void applyTo(int value, StartParameterInternal settings, Origin origin) {
+            settings.setConfigurationCacheMaxProblems(value);
+        }
+    }
+
+    public static class ConfigurationCacheRecreateOption extends EnabledOnlyBooleanBuildOption<StartParameterInternal> {
+
+        public static final String PROPERTY_NAME = "org.gradle.unsafe.configuration-cache.recreate-cache";
+        public static final String LONG_OPTION = "configuration-cache-recreate-cache";
```

Comment:
This could be changed to `--recreate-configuration-cache`

Reply:
or `--reconfigure`

Reply:
I wouldn't add a command-line option until we know we need one, based on feedback.

Reply:
Command line option removed, gradle and system property kept